Fix component cache for polymorphic field re-renders

tintinthong opened the pull request:
The point of this PR is to enable polymorphic fields to re-render upon assigning a subclass to a field of baseclass type. Previously, although the JSON is correctly serialised, the templates were not re-rendering

<img width="998" alt="Screenshot 2025-02-24 at 16 03 18" src="https://github.com/user-attachments/assets/b21a1a74-c966-4c01-8021-5c3ae02db4da" />

<img width="881" alt="Screenshot 2025-02-24 at 16 03 30" src="https://github.com/user-attachments/assets/a884a764-bc45-44d4-ad0f-abe44f3613e2" />

**Solution**

The main issue is that `getBoxComponent` was serving a stale cache (`componentCache`) bcos it did not take into account changes in the card or field class ie `cardOrField`. So we added a value inside the cache storing `cardOrField` which then serves the most up-to-date BoxComponent if `cardOrField` changed.

**Digression Risk**

~~As part of pairing on Thursday with @ef4, the outer cache at `fieldsComponentsFor` (`stableComponents`) was already serving a stale cache that was preventing the code from returning the most up-to-date cache in boxComponent. This several layers of caching made it difficult to debug. We initially thought that this outer-cache was not needed.~~

~~BUT afterwards I found out that the cache was made for the purpose preventing eager re-renders that caused `linksToManyComponent` and `containsManyComponent` to re-render -- so I kept the cache in place. Here is the relevant PR https://github.com/cardstack/boxel/pull/860 by @jurgenwerk. I eventually, decided to keep the cache for specific fields ie plural fields like `linksToManyComponent` and `containsManyComponent` for simplicity.. also to unblock Spec and Playground work.~~

~~I plan to pair further with @ef4 to find the best solution here.. I think the more general solution wud be to allow `getBoxComponent` to have better handling for plurals.~~

tintinthong committed:
remove outer cache in proxy. Update inner cache for box component to take into account when cardOrField changes

tintinthong committed:
add re-rendering test for polymorphic field

tintinthong committed:
Re-add stable cache in fields proxy but only for plural components

tintinthong committed:
Add polymorphic filed example

burieberry commented:
Is the template not rerendering when clicking the button still an issue? I tried on latest main and didn't have any issues.

tintinthong commented:
> Is the template not rerendering when clicking the button still an issue? I tried on latest main and didn't have any issues.

wdym. Where did u see that there was a re-render of the field? Particularly, when ur `somePersonField=contains(FieldDef)`

tintinthong commented:
> Is the template not rerendering when clicking the button still an issue? I tried on latest main and didn't have any issues.

I tried this test on main. I dont think it works

<img width="1040" alt="Screenshot 2025-02-27 at 02 46 23" src="https://github.com/user-attachments/assets/c21e0a62-a438-417a-95ac-e69952b729fa" />

burieberry commented:
> > Is the template not rerendering when clicking the button still an issue? I tried on latest main and didn't have any issues.
> 
> I tried this test on main. I dont think it works
> 

Sorry ok I was able to reproduce the issue. Previously didn't see it because I was switching between formats on code mode preview. Confirm that this PR resolves it.

tintinthong committed:
move cache to linksToMany

tintinthong committed:
fix for undefined boxed element that break many many tests when using linksToMany cache

tintinthong committed:
missing changes to linksToManyComponent cache

tintinthong committed:
Add test for polymorphic field within a card that is linked to another

tintinthong committed:
add more example

tintinthong commented:
I think this is ready for review

The continuation of this PR. I have removed the fields proxy cache and moved the cache into the linksToManyComponent. containsMany and linksToMany examples should still have the stability they have had previously.

tintinthong commented:
Its great that the change in cache introduced inside getBoxComponent (that makes cardOrField as a dependency) here fixes the re-ordering template issue seen here https://github.com/cardstack/boxel/pull/2241. 

Essentially, it invalidates the componentCache of the box component but that also means any component state is destroyed (without auto-save).

ef4 reviewed (approved):
We think this is a reasonable step forward without taking on a big refactor. 

It's hard to be sure this doesn't change component stability somewhere else, but we haven't been able to find any examples in the tests. So we can move ahead with this and folks should be on the lookout for any component state loss regressions that we can debug.

tintinthong committed:
remove cardOrField from linksToManyComponentCache

tintinthong committed:
Merge remote-tracking branch 'origin/main' into fix-component-cache-for-field
